src/compiler/clang.rs

```diff
@@ -57,49 +58,69 @@ pub fn compile<T>(creator: &T,
                   -> Box<Future<Item=(Cacheable, process::Output), Error=io::Error>>
     where T: CommandCreatorSync,
 {
-    let mut creator = creator.clone();
-    let compiler = compiler.clone();
-    let parsed_args = parsed_args.clone();
-    let cwd = cwd.to_string();
-
-    pool.spawn_fn(move || {
-        trace!("compile");
-        // Clang needs a temporary file for compilation, otherwise debug info
-        // doesn't have a reference to the input file.
+    trace!("compile");
+    // Clang needs a temporary file for compilation, otherwise debug info
+    // doesn't have a reference to the input file.
+    let input = parsed_args.input.clone();
+    let write = pool.spawn_fn(move || {
```

Comment:
Excellent idea! I count 4 locations which had to independently remember "oh don't forget to *not* drop `TempDir` early"!